Question 1: Part A

In [54]:
import pandas as pd
import numpy as np

In [6]:
dsf = pd.read_csv(r'C:\Users\PhD\Documents\2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv')

In [8]:
dsf.order_amount.mean()

3145.128

As the mean function by pandas above shows, the naive calculation of $3145.13 clearly came from just dividing the sum of all order values by the number of order values, without clearing out huge outliers that shifted it high up.

In [9]:
dsf.order_amount.std()

41282.539348788196

In [13]:
dsf.order_amount.max() - dsf.order_amount.min()

703910

As the function above shows, this dataset has an incredibly large standard deviation at $41282.54. The mean function is nowehere near an accurate description of the average order value as the data values are very dispersed with respect to the mean. Also the range of the dataset, $703910, doesn't seem like a normal value considering there are only 5000 values, showing us how impactful the outliers in the data are with respect to the mean.

Finding the number of entries for different order amounts will provide a good overview of the exact number of outliers and their values

In [26]:
diff_order_amounts = dsf.groupby(['order_amount']).size().reset_index(name='count').sort_values(by='order_amount', ascending=False)

In [29]:
diff_order_amounts.head(15)

order_amount  count
257        704000     17
256        154350      1
255        102900      1
254         77175      9
253         51450     16
252         25725     19
251          1760      1
250          1408      2
249          1086      1
248          1064      1
247          1056      3
246           980      1
245           965      1
244           960      2
243           948      1

The fact that there are 17 entries for $704000 shows that this dataset is very skewed and there are a lot of outliers. To detect and exclude these outliers, we can take an approach calculating the Z-Score of each value in the order_amount column, then taking the absolute Z-Score of the column, and deleting all values with a zscore above 0.5 or below -0.5.

In [38]:
dsf['zscore']=(dsf.order_amount - dsf.order_amount.mean())/dsf.order_amount.std(ddof=0)
dsf.head(5)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at    zscore  
0  2017-03-13 12:36:56 -0.070766  
1  2017-03-03 17:38:52 -0.074013  
2   2017-03-14 4:23:56 -0.072705  
3  2017-03-26 12:43:37 -0.072414  
4   2017-03-01 4:35:11 -0.072414

In [50]:
dsf_standardized = dsf[(dsf.zscore>-0.5)&(dsf.zscore<0.5)]
dsf_standardized.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at    zscore  
0  2017-03-13 12:36:56 -0.070766  
1  2017-03-03 17:38:52 -0.074013  
2   2017-03-14 4:23:56 -0.072705  
3  2017-03-26 12:43:37 -0.072414  
4   2017-03-01 4:35:11 -0.072414

In [51]:
dsf.shape[0]-dsf_standardized.shape[0]

63

We will probably get a much more believable value for an AOV with this new dataframe with removed outliers.

Question 1: Part B

I would report the mean value for the new dataframe with outliers removed, summing and dividing.

Question 1: Part C

In [52]:
dsf_standardized.order_amount.describe()

count    4937.000000
mean      302.580514
std       160.804912
min        90.000000
25%       163.000000
50%       284.000000
75%       387.000000
max      1760.000000
Name: order_amount, dtype: float64

The mean value comes out at $302 with a standard deviation of $160, a much better result with better data.

-------------------------------------------------------------------------------------------------------------------------------

Question 2: Part A

SELECT Shippers.ShipperName, COUNT(Orders.OrderID) AS NumberOfOrders FROM Orders
LEFT JOIN Shippers ON Orders.ShipperID
GROUP BY ShipperName;

This will give a table showing the number of orders shipped by each of the 3 shippers. Speedy Express shipped 54 orders.

Question 2: Part B

Select Employees.LastName, Count(Orders.OrderID) As NumberOfOrders
From Orders
Left Join Employees ON Orders.EmployeeID = Employees.EmployeeID
Group by LastName Order by NumberOfOrders Desc

This will give a table of all employees and the number of orders associated with them. According to the table, the last name of the employee with most orders is Peacock.

Question 2: Part C

Select Products.ProductName, Customers.Country
Count(Orders.OrderID) as NumberOfOrders FROM Customers
Right Join Orders on Customer.CustomerID = Orders.CustomerID
Right Join Order_details on Order.OrderID = Order_details.OrderID
Right Join Products on Order_details.ProductID = Products.ProductID
Where Customers.Country = 'Germany'
Group by ProductName 
Order by NumberOfOrders Desc

This will give a table of all products ordered in Germany with Number Of Orders listed in descending order. 
Gorgonzola Telino was the product most ordered by customers in Germany.